### PR TITLE
Adds cancer stats stratification data, filtering options

### PR DESCRIPTION
**Pull request by falquaddoomi**
This PR introduces two separate, but related improvements:
- updates the CancerStats model to include "RE" (race/ethnicity) and "Sex" fields that are now included in CancerInFocus dataset; the newest database dump, as of Feb 22nd 2024, includes these new columns
- adds support for "factors", additional categorical-valued columns in the dataset on which we can subset the data

Addresses #30, but we probably want to include the frontend part before we consider it closed. Feel free to use this PR for the frontend changes, or open a new one that references this one if you prefer.

## Terminology: "Factors"

*First: I'm of course open to debating the terminology, so if the terms don't make sense to you, please tell me!*

The name "factors" comes from R's "factor" datatype; in R, factors are used as an efficient way to represent values that take on a discrete set of categorical values. For example, a "Color" column that only takes on the values "Red", "Green", and "Blue" would be a candidate for a factor, and in R's case, each of these possible values would be called a "level". I've chosen to call them "values" of the factor rather than "levels" just to keep things simpler for us.

### The Data

Tables in CancerInFocus are organized first by geometry (county or tract), then by category, then by measure; each measure has an associated FIPS ID (i.e., identifying a county or tract) and a value for that FIPS. With the introduction of these additional "RE" and "Sex" columns (aka factors), each cancer-related table now has a row for the combination of the possible values for all the factors for each FIPS. For example, in the cancer mortality data, Denver County previously had a single row for "All Sites". Now, there is a row for (RE=All, Sex=All), a row for (RE="White NH", Sex=All), a row for RE="White NH", Sex=Male, etc.

A consequence of this design is that one must always filter the dataset by its factors to get a correct response. For example, to get back the old non-stratified dataset you must query the tables with RE=All and Sex=All. The API generally handles that for you; if you don't specify specific values for factors, they're assumed to be the factors' default.

Currently, the "RE" field takes on the values "All" (the default), "White NH", "Black NH", and "Hispanic". The "Sex" field currently takes on the values "All" (the default), "Female", and "Male". Despite these being the current values, this PR's implementation only hardcodes the default value and human-readable labels for the currently-observed values (e.g., "White NH" is labeled "White, Non-Hispanic" in the metadata). The idea behind all this is that the existing factors will continue to work even if new versions of the data introduce new factors or new values for existing factors. We can catch up to extending the factors and adding human-readable labels as we notice issues without things breaking completely.

## Accessing Factor Metadata

This PR changes the `/stats/measures/` endpoint to also return information about factors for each table, if available. Under each category, in addition to the `measures` field, there is now a `factors` field which is `{}` for most tables, but contains factor information for tables that have factors (e.g., the cancer incidence and mortality county-level tables).  The new response for, e.g., cancer mortality, looks like this (with measures left out for brevity):
```json
{
    "county": {
        "label": "County",
        "categories": {
            "cancermortality": {
                "label": "Cancer Mortality (age-adj per 100k)",
                "measures": { },
                "factors": {
                    "RE": {
                        "label": "Race/Ethnicity",
                        "default": "All",
                        "values": {
                            "All": "All",
                            "Black NH": "Black (non-Hispanic)",
                            "Hispanic": "Hispanic",
                            "White NH": "White (Non-Hispanic)"
                        }
                    },
                    "Sex": {
                        "label": "Sex",
                        "default": "All",
                        "values": {
                            "All": "All",
                            "Female": "Female",
                            "Male": "Male"
                        }
                    }
                }
            }
        }
    },
    "tract": { }
}
```
The keys in `factors` are the identifiers you should use when interacting with the API. Under each entry in factors, you'll find:
- `label`: the human-readable version of the factor, for display in the web UI
- `default`: a value representing "no selection" in the data; should be the default selection in the web UI. (It's also used in the API when querying a table that has factors but when no specific value for a factor is given.)
- `values`: a dictionary of value-label pairs for the column.
  - The values are computed from the table, and like the factor keys should be used when interacting with the API.
  - The labels come from the `backend.app.models.cif.FACTOR_DESCRIPTIONS` dictionary; if no match is found there, the value is used as-is as the label. These labels are intended to be used as the labels for the factors' dropdown in the web UI.

### Using Factors to Subset

The `/stats/<type>/<model>/fips-value` endpoint now takes an optional querystring argument, `filters`, which can be used for subsetting a table based on its factors. The argument takes a semicolon-delimited list of comma-delimited `<factor>:<value>` pairs. For example, the string "RE:White NH;Sex:Female" would filter the dataset to only return rows where the "RE" column is "White NH" and the "Sex" column is "Female". If a factor is left out, it takes on its default value as specified by the "default" field in the metadata for that factor.

Other notes:
- White space is trimmed at each level of parsing, so the string "  RE : White NH  ; Sex :   Female   " is equivalent to "RE:White NH;Sex:Female". (White space is retained within a key or value).
- As of now there are no values that include one of our separators (":", or ";"), but if that should occur we might move toward encoding the filter string as, say, JSON.

### CODE REVIEW

**Comment by vincerubinetti** [comment]
Just getting to this now. Read through your description. The terminology and approach all make sense to me.

> As of now there are no values that include one of our separators (":", or ";"), but if that should occur we might move toward encoding the filter string as, say, JSON.

If this becomes the case we might want to just enforce some url/filepath-safe identifier, because we'd maybe want a non-obfustracted human-readable way to put selected factors in the UI urls for users to share.

I'll probably start doing the frontend changes in this PR and that will be my review of testing to make sure everything works. That'll probably be better than me looking over your code, though I'll still glance through it.

File: backend/app/routers/statistics.py [hunk 1]
```diff
@@ -52,9 +55,15 @@ class FIPSMeasureResponse(BaseModel):
 class MeasuresMetaResponse(BaseModel):
     label: str
 
+class FactorMetaResponse(BaseModel):
+    label : str
+    default : str | None
+    values : dict[str, str]
+
 class CateogryMetaResponse(BaseModel):
```
Comment (vincerubinetti) [inline backend/app/routers/statistics.py:63]: Just noticed typo here

Comment (vincerubinetti) [inline backend/app/routers/statistics.py:63]: Just noticed typo here

**Comment by falquaddoomi** [comment]
> If this becomes the case we might want to just enforce some url/filepath-safe identifier, because we'd maybe want a non-obfustracted human-readable way to put selected factors in the UI urls for users to share.

That's a good point about readability; I doubt being unable to use semicolons and colons in the factor IDs or values is really going to impact readability, but switching to a different format would, so just let's stick with our current format. It's not hard to strip out the colons and semicolons during both data ingest and querying.

> I'll probably start doing the frontend changes in this PR and that will be my review of testing to make sure everything works. That'll probably be better than me looking over your code, though I'll still glance through it.

Sounds good. I'd give running the stack locally a try while you develop the frontend, assuming you're not already doing that.  If you'd like me to deploy it to prod I can do that, too; there shouldn't be any changes that would break the older frontend.

**Review by vincerubinetti — APPROVED** [review]

**Pull request merged**